specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2017-10-01/containerregistry.json

```diff
@@ -1497,6 +1497,10 @@
         "storageAccount": {
           "$ref": "#/definitions/StorageAccountProperties",
           "description": "The properties of the storage account for the container registry. Only applicable to Classic SKU."
+        },
```

Comment:
What is the service behavior if a client performs a PUT with this API version and specifies networkRuleset, and another client GETs & PUTs the resource with a lower API version? Is the downlevel PUT blocked, or do you ignore networkRuleset for the update?

Reply:
Lower versions ignore `networkRuleset` in PUT/PATCH, also won't return it in GET.

Reply:
Seems reasonable to me. @KrisBash - signing off on your behalf.

Reply:
Apologies, did not check thouroughly. @djyou  - seems like this new property is added in the same api-version. If your service were doing the right PUT behavior (replace semantics) then this would be a breaking change. I would recommend doing this in a new API-version.

Reply:
Suppose I created a registry using this api-version `2010-10-01` with the new property `networkRuleSet`, I'd have the following json (other irrelevant properties ignored):
```
{
    "name": "myRegistry",
    "networkRuleSet": {
        "defaultAction": "Deny",
        "virtualNetworkRules": [
            {
                "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myRG/providers/Microsoft.Network/virtualNetworks/myvnet/subnets/default"
            }
        ]
    }
}
```

Now if I do a PUT using lower api-versions, since they don't support the property `networkRuleSet`, it will be ignored (the same as whatever extra properties passed in). So:

PUT request:
```
{
    "name": "myRegistry",
    "networkRuleSet": {
        "defaultAction": "Deny",
        "virtualNetworkRules": [
            {
                "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myRG/providers/Microsoft.Network/virtualNetworks/myvnet/subnets/default"
            }
        ]
    },
    "otherPropertyIgnored": "value"
}
```

PUT/GET response:
```
{
    "name": "myRegistry"
}
```

However, GET using this api-version `2017-10-01` will return the full json:
```
{
    "name": "myRegistry",
    "networkRuleSet": {
        "defaultAction": "Deny",
        "virtualNetworkRules": [
            {
                "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myRG/providers/Microsoft.Network/virtualNetworks/myvnet/subnets/default"
            }
        ]
    }
}
```

@ravbhatnagar Could you clarify which part is a breaking change?

Reply:
Within the same api-version `2017-10-01`, if I created a registry with the new property `networkRuleSet`:
```
{
    "name": "myRegistry",
    "networkRuleSet": {
        "defaultAction": "Deny",
        "virtualNetworkRules": [
            {
                "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myRG/providers/Microsoft.Network/virtualNetworks/myvnet/subnets/default"
            }
        ]
    }
}
```

Now if I do a PUT using old clients on the same api-version `2017-10-01`, since they don't have the new property `networkRuleSet`, it's essentially a `null` value:

PUT request:
```
{
    "name": "myRegistry",
    "networkRuleSet": null
}
```

On the server side, since the resource `myRegistry` already exists, it enters the update path, which replaces whatever values provided in the request, while ignoring unprovided ones.

PUT response:
```
{
    "name": "myRegistry",
    "networkRuleSet": {
        "defaultAction": "Deny",
        "virtualNetworkRules": [
            {
                "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myRG/providers/Microsoft.Network/virtualNetworks/myvnet/subnets/default"
            }
        ]
    }
}
```

Basically, this new property behaves in a same way as existing properties, such as `adminUserEnabled`, if it has a value, and a second PUT doesn't provide a value, we don't clear it's value.

Reply:
Yes, this is exactly what I meant. Currently your RP does not implement PUT in the way it should have been where properties missing in the request are wiped out from the resource on the server. So in essence, following the "replace" semantics for PUT. Since your RP is not following the right model, it is currently not a breaking change. But for most services (specially the newer ones) in Azure, such changes go in a new api-version and PUT is a replace.

Reply:
Thank you @ravbhatnagar We had an offline discussion within the team, since historically this resource didn't follow the replace semantic and introducing the right pattern itself is a breaking change (due to the existing properties `adminUserEnabled`, `storageAccountProperties`, etc.), do you think we can have this merged in for this api-verison, and we will introduce a new api-version that follows the right pattern? /cc @mnltejaswini @sajayantony

Reply:
@ravbhatnagar We queued up an item to redesign the resource APIS with a newer api version and  other new features. As part of it, we can update the APIs to implement the semantics correctly.

Reply:
Sounds good. Signing off from ARM side. opened an issue to track it.

Reply:
Thank you!